[inductor][subgraph] Plumbing to get ShapeAsConstantBuffer from subgraph to main graph output

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139325
* __->__ #147559


I am unable to create a test case that fails without the next PR. The idea is to have a symint which is returned by the inner subgraph and then returned by the forward graph after partitioning.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

anijain2305 committed:
[inductor][subgraph] Plumbing to get ShapeAsConstantBuffer from subgraph to main graph output

[ghstack-poisoned]

eellison commented on torch/_inductor/ir.py:
```diff
@@ -5128,6 +5133,11 @@ def realize_input(cls, x):  # type: ignore[no-untyped-def]
                 except NotImplementedError:
                     pass
         if isinstance(x, StorageBox):
+            if isinstance(x.data, ShapeAsConstantBuffer):
```
do you know why we're mapping ShapeAsConstantBuffer with StorageBox ? that seems weird

anijain2305 replied:
It happens here. I can prevent wrapping of symints here. Is that more intuitive?

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L6779

eellison replied:
Yes, I think so.

anijain2305 replied:
Ok, updated.

anijain2305 committed:
Update on "[inductor][subgraph] Plumbing to get ShapeAsConstantBuffer from subgraph to main graph output"



I am unable to create a test case that fails without the next PR. The idea is to have a symint which is returned by the inner subgraph and then returned by the forward graph after partitioning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][subgraph] Plumbing to get ShapeAsConstantBuffer from subgraph to main graph output"



I am unable to create a test case that fails without the next PR. The idea is to have a symint which is returned by the inner subgraph and then returned by the forward graph after partitioning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][subgraph] Plumbing to get ShapeAsConstantBuffer from subgraph to main graph output"



I am unable to create a test case that fails without the next PR. The idea is to have a symint which is returned by the inner subgraph and then returned by the forward graph after partitioning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

anijain2305 commented:
@eellison This should be ready now.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -7067,6 +7068,8 @@ def __str__(self) -> str:
 class TensorBox(MutableBox):
     @staticmethod
     def create(data):  # type: ignore[no-untyped-def]
+        if isinstance(data, ShapeAsConstantBuffer):
```
do you know when we create these ? just curious

anijain2305 replied:
I think in `realize_input` of ExternKernel. If that input happens to be output as well, then this PR is needed to make it work.

anijain2305 commented:
@pytorchbot merge -i
